### ISSUE TITLE
Suggestion for ComboBox::show_ui()'s usability 

### ISSUE THREAD

**Issue by sowbug** [emilk/egui#3417]
Most egui things return a `Response`, enabling generic behavior like this: `if ui.add(something()).changed() { deal_with_it(); }`. But `ComboBox`'s main method, `show_ui()`, has an atypical signature:

```rust
pub fn show_ui<R>(
    self,
    ui: &mut Ui,
    menu_contents: impl FnOnce(&mut Ui) -> R
) -> InnerResponse<Option<R>>
```

As far as I can tell, this leads to code that either returns the `Response` from `ui.selectable_value()` when the user picks that option in the dropdown, or... has to synthesize a `Response` because none of the select options were picked.

Example (also at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6fdd1cadfa4f40f897ecc01546fa65d3):

```rust
#[derive(IntoStaticStr, EnumIter, PartialEq, Display, Copy, Clone)]
enum Fruit {
    Apple,
    Banana,
}

struct MyStruct {
    fruit: Fruit,
}

impl MyStruct {
    // This method signature can't be changed because it's actually a trait
    // method in my real code
    fn ui(&mut self, ui: &mut eframe::egui::Ui) -> eframe::egui::Response {
        let r = ComboBox::from_label("Fruit")
            .selected_text(self.fruit.to_string())
            .show_ui(ui, |ui| {
                for fruit in Fruit::iter() {
                    let fruit_str: &'static str = fruit.into();
                    let response = ui.selectable_value(&mut self.fruit, fruit, fruit_str);
                    if response.changed() {
                        return response;
                    }
                }
                ui.label("This label exists only to produce a Response") // UGH
            });
        if let Some(inner) = r.inner {
            return inner;
        }
        r.response
    }
}
```

That's silly; this one is better, but still cumbersome (also at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f3535f411156bd6b167f9bf0ab31ab52):

```rust
#[derive(IntoStaticStr, EnumIter, PartialEq, Display, Copy, Clone)]
enum Fruit {
    Apple,
    Banana,
}

struct MyStruct {
    fruit: Fruit,
}

impl MyStruct {
    // This method signature can't be changed because it's actually a trait
    // method in the real code
    fn ui(&mut self, ui: &mut eframe::egui::Ui) -> eframe::egui::Response {
        let mut r = ComboBox::from_label("Fruit")
            .selected_text(self.fruit.to_string())
            .show_ui(ui, |ui| {
                let mut bool_response = false;
                for fruit in Fruit::iter() {
                    let fruit_str: &'static str = fruit.into();
                    if ui
                        .selectable_value(&mut self.fruit, fruit, fruit_str)
                        .changed()
                    {
                        bool_response = true;
                    }
                }
                bool_response
            });
        if let Some(inner) = r.inner {
            if inner {
                r.response.mark_changed(); // We're stealing ComboBox's Response just to send the changed flag
            }
        }
        r.response
    }
}
```

Maybe this is unavoidable because ComboBox is a complicated widget. I think it would be more intuitive to use if there were a method the inner code could call to set ComboBox's changed() flag to true (which is basically what I'm doing in my second code snippet).

Sorry if I'm just confused about something basic.

**Comment by lwiklendt** [comment]
You can use [`Response.union`](https://docs.rs/egui/latest/egui/struct.Response.html#method.union) to combine the `Response` from each `ui.selectable_value`, say, via [`reduce`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.reduce) on the `Fruit::iter()`, and finnaly do a `.inner.is_some_and(|r| r.changed())` on the `InnerResponse` to get if a change occured.

**Comment by sowbug** [comment]
Thanks, @lwiklendt, that unblocked me. I haven't yet run this code (it's not too late where I am to write code, but far too late to debug it :) ), but at least it builds:

```rust
        let r = ComboBox::new(ui.next_auto_id(), "Fruit")
            .show_ui(ui, |ui| {
                Fruit::iter()
                    .map(|f| {
                        ui.selectable_value(self.inner, f, f.to_string())
                    })
                    .reduce(|acc, r| acc | r)
                    .unwrap()
            });
        let response = r.inner.unwrap_or(r.response);
```

... or something like that.

**Comment by sowbug** [comment]
I've now used this approach in a couple places in my code, and it's satisfactory. I still think the API could be more intuitive, but I don't have a proposal.

The sample code I provided above will panic if the iterator is empty. That isn't possible in the case of a `strum::EnumIter`, but it is if the iterator is dynamic. In that case, I would return the `reduce` result to `show_ui()`, then handle the `InnerResponse<Option<Option<...>>>` appropriately.
